rero_ils/modules/loans/extensions.py

```diff
@@ -18,21 +18,61 @@
 """RERO ILS common record extensions."""
 
 
-from datetime import timedelta, timezone
+from datetime import datetime, timedelta, timezone
 
 import ciso8601
+from flask import current_app
 from invenio_records.extensions import RecordExtension
 
+from rero_ils.modules.libraries.exceptions import LibraryNeverOpen
+from rero_ils.modules.loans.models import LoanState
 
-class SetDueSoonDate(RecordExtension):
-    """Set the due soon date when the loan is updated."""
 
-    def pre_commit(self, record):
-        """Add the due soon date.
+class CirculationDatesExtension(RecordExtension):
+    """Add some dates to manage circulation operation."""
+
+    @staticmethod
+    def _add_request_expiration_date(record):
+        """Add the request expiration date to record is needed.
+
+        :param record: the record metadata.
+        """
+        from .utils import get_circ_policy
+        if record.state == LoanState.ITEM_AT_DESK and \
+           'request_expire_date' not in record:
+            cipo = get_circ_policy(record)
+            request_duration = cipo.get('request_duration')
+            library = record.pickup_library
+            if cipo.get('allow_requests') and request_duration and library:
+                # the expiration date should be calculated using the pickup
+                # library calendar
+                trans_date = ciso8601.parse_datetime(record.transaction_date)
+                try:
+                    expire_date = trans_date + timedelta(days=request_duration)
+                    expire_date = expire_date.replace(
+                        hour=23, minute=59, second=00, microsecond=000,
+                        tzinfo=library.get_timezone()
+                    )
+                    expire_date = library.next_open(expire_date)
+                except LibraryNeverOpen:
+                    # 10 days by default ... it's better than place a random
+                    # date value
+                    default_duration = current_app.config.get(
+                        'RERO_ILS_DEFAULT_REQUEST_DURATION', 10)
+                    expire_date = trans_date + timedelta(days=default_duration)
+                    expire_date = expire_date.replace(
+                        hour=23, minute=59, second=00, microsecond=000,
+                        tzinfo=library.get_timezone()
+                    )
+                record['request_expire_date'] = expire_date.isoformat()
```

Comment:
by careful here `request_expire_date` in the `loan` resource is to manage the expiration globally for the any loan [`PENDING`, `ITEM_AT_DESK`].

Reply:
I think all what you need to do is to calculate the `expired validated requests` as follows:
for all loans at desk:
 if transaction_date + cipo.get('keep_until_parameter') > sysdate:
         cancel loan